### PR TITLE
Fixes #32608 - Add variable overrides to host detail

### PR DESCRIPTION
**Pull request by xprazak2**
We might need to think about a best way how to display composite values such as yaml or json in the future. Just showing the raw value for now.

vars

### CODE REVIEW

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTableHelper.js
Comment (amirfefer) [inline webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTableHelper.js:19]: does it make sense to store the data in the redux's store?

**Review by amirfefer — CHANGES_REQUESTED** [review]
Thanks @xprazak2 👍 
I left a few inline comments.

File: webpack/components/AnsibleHostDetail/AnsibleHostDetail.js
Comment (xprazak2) [reply webpack/components/AnsibleHostDetail/AnsibleHostDetail.js:5]: I defer to @MariSvirik for a final decision...

Comment (MariSvirik) [reply webpack/components/AnsibleHostDetail/AnsibleHostDetail.js:5]: @xprazak2 as we know how the populated page will look like, we should use skeleton. https://www.patternfly.org/v4/components/skeleton/design-guidelines

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTableHelper.js
Comment (xprazak2) [reply webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTableHelper.js:19]: Doing that would not bring any benefits.

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverrides.js
Comment (xprazak2) [reply webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverrides.js:24]: Yes, and it counts on specific constants to be present, which is not the case here.

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverrides.js
Comment (xprazak2) [reply webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverrides.js]: Apollo takes care of data, so no action required here. I extracted the function out of the component.

File: package.json
Comment (xprazak2) [reply package.json]: Yes, removed.

File: webpack/components/AnsibleHostDetail/AnsibleHostDetail.js
Comment (xprazak2) [reply webpack/components/AnsibleHostDetail/AnsibleHostDetail.js:5]: Thanks, I replaced loader with skeletons.

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTable.js
Comment (amirfefer) [inline webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTable.js:37]: There are two components at least with the same key due to using `index` as a key.
[react docs](https://reactjs.org/docs/lists-and-keys.html) doesn't recommend using `index` as a key because it does not represent the data correctly and therefore can break dynamic data.

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTable.js
Comment (amirfefer) [inline webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTable.js:39]: looks like a good candidate for the component key

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTableHelper.js
Comment (amirfefer) [reply webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTableHelper.js:19]: I mean, the data comes from the server, like we have the API middleware and selectors for fetching, storing, cleaning, formatting, and manipulate data. I wrote a suggestion in another comment.

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTableHelper.js
Comment (amirfefer) [inline webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTableHelper.js:25]: if the user can't edit it, we can make it disabled rather than removing it?
@MariSvirik any thoughts?

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTableHelper.js
Comment (amirfefer) [inline webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTableHelper.js:45]: switch case? to me it more readable and easy to extend / change

```js
switch (paramterType)
{
    case "boolean":
     return  value ? <CheckIcon /> : <TimesIcon />;
    case "yaml":
    case "hash":
    case "array":
      return JSON.stringify(value);
    default:
       return value
}
```

**Review by amirfefer — CHANGES_REQUESTED** [review]
Thanks @xprazak2 !
I left a few inline comments

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTableHelper.js
Comment (xprazak2) [reply webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTableHelper.js:25]: I return value without link 3 lines below.

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesHelper.js
Comment (xprazak2) [reply webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesHelper.js]: Changed, though this is the data we requested in query so if there is no `role.ansibleVariablesWithOverrides.nodes`, it means we do not wait for request to resolve or we have not handled error response properly.

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverrides.test.js
Comment (xprazak2) [reply webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverrides.test.js:143]: It does not seem to work for me.

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverrides.js
Comment (xprazak2) [reply webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverrides.js:18]: With the hook handling the mapping to status constants, the component looks like the following:

```js
const AnsibleVariableOverrides = ({ hostAttrs }) => {
  const { id } = hostAttrs;

  const { status, data, error } = useQueryWithStatus({ hostId: id, id: encodeId('Host', id) });

  const errorNode = (<div>{error ? error.message : ''}</div>)

  return (
    <SkeletonLoader status={status} errorNode={errorNode}>
      <div className="ansible-tab-margin">
        <AnsibleVariableOverridesTable
          variables={reorderVariables(data ? data.host.allAnsibleRoles.nodes : [])}
          hostAttrs={hostAttrs}
        />
      </div>
    </SkeletonLoader>
  );
```
So I still need to check for errors and data presence in component, but now I use additional hook wrapper and status mapping - I added more things only to get what I already had. I kept it as was for now but I am willing to change if you really think the version with wrapper is better. 

> Also, with this suggested custom hook, the data can cam be stored in foreman's redux store (like useAPI), so data manipulation can be done via selectors and memorized by reselect if needed

Yes, and it would mean having the same data in 2 places - redux store and apollo cache. That would effectively neutralize some of apollo's strong features, such as [auto-updates](https://www.apollographql.com/docs/react/data/mutations/#updating-a-single-existing-entity) after mutation, because we would need to manually update the store to match the data from apollo.

I am strongly against putting any data fetched by apollo into redux store.

File: package.json
Comment (xprazak2) [reply package.json:16]: Not yet: https://github.com/theforeman/foreman-js/pull/320

**Comment by xprazak2** [comment]
Thank you for your review @amirfefer, updated.

File: webpack/components/AnsibleHostDetail/AnsibleHostDetail.js
Comment (tbrisker) [inline webpack/components/AnsibleHostDetail/AnsibleHostDetail.js:24]: This won't show the loading state as there are children

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTableHelper.js
Comment (tbrisker) [inline webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTableHelper.js:3]: These imports won't work at build time, you need to import them from @paternfly/react-icons

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTable.js
Comment (tbrisker) [inline webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTable.js:21]: Should the role be the first column and group the variables from the same role together?

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTableHelper.js
Comment (tbrisker) [inline webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTableHelper.js:12]: Why only these three and not other resources?

File: app/graphql/overriden_ansible_variable_presenter.rb
Comment (tbrisker) [inline app/graphql/overriden_ansible_variable_presenter.rb:23]:
```suggestion
    element.classify.safe_constantize
```
Would also prevent the need to disable the rubocop

File: app/graphql/types/ansible_variable_override.rb
Comment (tbrisker) [inline app/graphql/types/ansible_variable_override.rb:7]: These names are a bit confusing, should we rename them to something more descriptive such as `override_source` or `overriding_resource`?

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTableHelper.js
Comment (tbrisker) [inline webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTableHelper.js:8]: Should the element be translated?

**Review by tbrisker — CHANGES_REQUESTED** [review]
Added my still relevant comments from reviewing the other PR, didn't give this PR a full review though.

File: app/graphql/types/ansible_variable_override.rb
Comment (xprazak2) [reply app/graphql/types/ansible_variable_override.rb:7]: These come directly from classification, are we sure we want to diverge on naming?

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTableHelper.js
Comment (xprazak2) [reply webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTableHelper.js:12]: These are the values offered in 'Prioritize attribute order' when editing. There are too many special cases that cannot be reliably handled in a generic way

1. Since matchers can be set up on virtually anything, there is no way to tell if we match on host attribute (such as comment) or on associated resource.
2. Some resources may may have naming mismatches similar to operating system
3. Katello has a very different page paths than Foreman and there is no way to tell where the resource comes from
4. Some resources extend host via facets

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTable.js
Comment (xprazak2) [reply webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTable.js:21]: No, we do not go for the same design as what we have in host form. The variables from the same role are grouped by default, I imagine we could have sortable header for role in the future.

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverrides.js
Comment (amirfefer) [reply webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverrides.js:18]: >  I am strongly against putting any data fetched by apollo into redux store.

I read about this a bit further  and I agree with you, we should use apollo as a [redux replacement](https://www.apollographql.com/blog/apollo-client/caching/dispatch-this-using-apollo-client-3-as-a-state-management-solution/)

**Comment by xprazak2** [comment]
@amirfefer, @tbrisker could you take another look please?

File: lib/foreman_ansible/register.rb
Comment (tbrisker) [inline lib/foreman_ansible/register.rb:102]: This seems to be causing the tests to fail with
```
ActiveRecord::StatementInvalid: PG::UndefinedTable: ERROR:  relation "ansible_roles" does not exist
LINE 8:  WHERE a.attrelid = '"ansible_roles"'::regclass
```
I guess it runs before the db is migrated so the table doesn't yet exit?

File: jest.config.js
Comment (tbrisker) [reply jest.config.js:10]: This can be dropped now that it's in `@theforeman/test`

File: package.json
Comment (tbrisker) [reply package.json:16]: this too

File: package.json
Comment (tbrisker) [reply package.json:26]: This as well, minimum version of `@theforeman/test` should be updated to 8.9.0

File: webpack/components/AnsibleHostDetail/AnsibleHostDetail.js
Comment (tbrisker) [inline webpack/components/AnsibleHostDetail/AnsibleHostDetail.js:16]: I think the skeleton loader isn't needed here, we can always show the tab itself, and the AnsibleVariableOverrides component already has a skeleton while it's loading

File: webpack/components/AnsibleHostDetail/AnsibleHostDetail.js
Comment (tbrisker) [inline webpack/components/AnsibleHostDetail/AnsibleHostDetail.js:23]: do we need two separate divs for these classes? could we combine them in one or even put the classnames on one of the other components (`Tab` or `AnsibleVariableOverrides`)?

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverrides.js
Comment (tbrisker) [inline webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverrides.js:29]: Here as well, could this be className attribute of the AnsibleVariableOverridesTable?

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesHelper.js
Comment (tbrisker) [inline webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesHelper.js:8]: Can we get the data from the server in the format we want it directly and avoid this complex mutation? IIUC all this changes is adding a roleName field on every variable, but I may be misunderstanding?

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTableHelper.js
Comment (tbrisker) [reply webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTableHelper.js:11]: > These are the values offered in 'Prioritize attribute order' when editing.

If I'm not mistaken, the default order is defined by a setting, and every role can have it's own override order. 
I agree matchers could be set on anything and it's hard to guess the correct path - perhaps as a first step we could drop the attempt to link to the resource and only display the type and name? Then we also don't need to check the `canEdit` permission

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/index.js
Comment (tbrisker) [inline webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/index.js:1]: I think in https://community.theforeman.org/t/rethinking-react-redux-folder-structure/24183 we agreed that the main component should be directly in the index.js file?

File: app/graphql/types/ansible_variable.rb
Comment (tbrisker) [inline app/graphql/types/ansible_variable.rb:17]: if we add 
```
field :ansible_role_name, String
```
I think we can avoid the complex mutation of the response

File: app/graphql/overriden_ansible_variable_presenter.rb
Comment (tbrisker) [inline app/graphql/overriden_ansible_variable_presenter.rb:34]: This can get quite slow for non-admin users if there are a lot of overrides

**Review by tbrisker — CHANGES_REQUESTED** [review]
Thanks @xprazak2, updated inline

File: webpack/components/AnsibleHostDetail/AnsibleHostDetail.js
Comment (xprazak2) [reply webpack/components/AnsibleHostDetail/AnsibleHostDetail.js:23]: I am not a css guru, but I do not think we can combine them since each sets a different bg color. I tried to pull on the inner one into the AnsibleVariableOverrides component, but that results in light gray skeleton on slightly darker gray background which does not look good.

File: webpack/components/AnsibleHostDetail/AnsibleHostDetail.js
Comment (xprazak2) [reply webpack/components/AnsibleHostDetail/AnsibleHostDetail.js:16]: The page makes request for the host attributes and the skeleton loader here waits for it to resolve so that it has host id available. The loader in AnsibleVariableOverrides waits for the variable values to be fetched.

**Comment by xprazak2** [comment]
@tbrisker, @amirfefer could you take another look please?

**Review by tbrisker — CHANGES_REQUESTED** [review]
Looking much cleaner now! just a few small comments inline

File: webpack/graphql/queries/variableOverrides.gql
Comment (tbrisker) [inline webpack/graphql/queries/variableOverrides.gql:4]: I might be missing some graphql background here, why do we need the extra `nodes` layer here and below?

**Review by tbrisker — COMMENTED** [review]
Looks good @xprazak2 ! only a couple of questions inline, and some linting issues that need resolving.

**Review by amirfefer — CHANGES_REQUESTED** [review]
Thanks @xprazak2 ! 
I left a few small comments, plus js tests fail

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/index.js
Comment (xprazak2) [reply webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/index.js]: True and we [wait for it](https://github.com/theforeman/foreman_ansible/blob/master/webpack/components/AnsibleHostDetail/AnsibleHostDetail.js#L18) to be present before the component is rendered.

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/index.js
Comment (xprazak2) [reply webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/index.js:15]: It might be used instead of `hostId`, but real id is still needed to construct a global id, so the whole component is rendered only after the id is available.

**Review by amirfefer — APPROVED** [review]
Thanks @xprazak2 !
LGTM

**Comment by tbrisker** [comment]
[test foreman_ansible]

File: webpack/graphql/queries/variableOverrides.gql
Comment (xprazak2) [reply webpack/graphql/queries/variableOverrides.gql:4]: It is because of Relay style connections that add additional information such as total count of the nodes.
```
allAnsibleRoles {
  totalCount
  pageInfo {
     hasNextPage
  }
  nodes {
   //...
  }
}
```

File: app/services/foreman_ansible/override_resolver.rb
Comment (xprazak2) [reply app/services/foreman_ansible/override_resolver.rb:7]: It is probably not an issue in a context of the host detail page, but the method relies on host not being nil.

Comment (tbrisker) [reply app/services/foreman_ansible/override_resolver.rb:7]: In case this is ever called with a `nil` host (e.g. in case we reuse it somewhere else), should we silently provide an empty hash or would it be better to throw an error?

File: webpack/graphql/queries/variableOverrides.gql
Comment (tbrisker) [reply webpack/graphql/queries/variableOverrides.gql:4]: Do we need this additional information here? if I'm not mistaken, all variables are loaded without pagination and we don't use the count anywhere.

Comment (tbrisker) [reply webpack/graphql/queries/variableOverrides.gql:4]: never mind, I played with it a bit and now I understand why this is needed.

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTableHelper.js
Comment (tbrisker) [reply webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesTableHelper.js:16]: I think that was the implementation in a previous iteration, but that won't work correctly for boolean values

File: webpack/globalIdHelper.js
Comment (tbrisker) [inline webpack/globalIdHelper.js:11]: Not crucial but I suppose that this should be provided by core rather than every plugin creating its own.

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesHelper.js
Comment (tbrisker) [inline webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesHelper.js:8]: We could use `flatMap` here which doesn't create an intermediate array for each iteration (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap#alternative) and is more readable:
```suggestion
roles.flatMap((role)=>(role.ansibleVariablesWithOverrides.nodes))
```
If I'm not mistaken, the conditional chaining operand is not needed since GraphQL guarantees that the response will always include these fields even if they are empty.

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesHelper.js
Comment (tbrisker) [inline webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesHelper.js:1]: The naming here is confusing IMHO, it took me a while to understand what this function actually does. 
We aren't reordering the variables, we are flattening the response, and the argument isn't the variables, it's the roles.
How about something like:
```suggestion
export const extractVariables = roles =>
```

**Review by tbrisker — CHANGES_REQUESTED** [review]

File: webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesHelper.js
Comment (xprazak2) [reply webpack/components/AnsibleHostDetail/components/AnsibleVariableOverrides/AnsibleVariableOverridesHelper.js]: Updated, conditional chaining is not needed as mentioned in [previous comment](https://github.com/theforeman/foreman_ansible/pull/420#discussion_r665386867)

File: webpack/globalIdHelper.js
Comment (xprazak2) [reply webpack/globalIdHelper.js:11]: Yes, this should move to core at some point.

**Review by tbrisker — APPROVED** [review]
Thanks @xprazak2 ! Great job pushing this through 🏅

File: app/services/foreman_ansible/override_resolver.rb
Comment (tbrisker) [inline app/services/foreman_ansible/override_resolver.rb:7]: rubocop isn't happy, i think this should be
```suggestion
      raise(Foreman::Exception.new('OverriderResolver needs a host to resolve overrides')) unless host
```

**Review by tbrisker — APPROVED** [review]
Thanks @xprazak2 !

**Pull request merged**
